### PR TITLE
When creating a session token, make sure it's the login password and …

### LINKED ISSUE #24730: if DAV clients log in via username and token, the session is killed
### Steps to reproduce
1. Generate a token
2. Configure your sync client with user:token
3. Start syncing

### Expected behaviour
The sync clients can authenticate successfully and a browser session is created (#24543 will fix that)

### Actual behaviour
The sync client can authenticate successfully, but as soon as the token is validated (can happen on login and on subsequent requests) the session is killed as the login password is wrong (when creating the session token the login password is actually a token).

The problematic code is here https://github.com/owncloud/core/blob/65eae43ab122bf1d04a5d6f8bee8e8b5ec439276/apps/dav/lib/Connector/Sabre/Auth.php#L106

A session token is created on successful client login. However, if the client uses a token instead of a password, we'd need to restore (decrypt) the login password first and then create the session token.

cc @LukasReschke

### PR DESCRIPTION
**Pull request by ChristophWurst**
…not a device token

fixes https://github.com/owncloud/core/issues/24730

### CODE REVIEW

File: lib/private/User/Session.php [hunk 1]
```diff
@@ -470,13 +470,24 @@ public function createSessionToken(IRequest $request, $uid, $loginName, $passwor
...
-
+			// No idea why this could happen, let's return false :see_no_evil:
+			return false;
 		}
+		$pwd = $this->getPassword($password);
+		$this->tokenProvider->generateToken($sessionId, $uid, $loginName, $pwd, $name);
 		return true;
 	}
 
+	private function getPassword($password) {
```
Comment (PVince81) [inline lib/private/User/Session.php]: PHPDoc ?

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -470,13 +470,24 @@ public function createSessionToken(IRequest $request, $uid, $loginName, $passwor
 		$name = isset($request->server['HTTP_USER_AGENT']) ? $request->server['HTTP_USER_AGENT'] : 'unknown browser';
 		try {
 			$sessionId = $this->session->getId();
-			$this->tokenProvider->generateToken($sessionId, $uid, $loginName, $password, $name);
 		} catch (SessionNotAvailableException $ex) {
-
+			// No idea why this could happen, let's return false :see_no_evil:
+			return false;
```
Comment (PVince81) [inline lib/private/User/Session.php]: Is it safe or should it throw further, especially if this is an exceptional case ?

**Comment by PVince81** [comment]
Unit tests would be nice :smile:

File: lib/private/User/Session.php (hunk 2 above)
Comment (ChristophWurst) [inline lib/private/User/Session.php]: Hm, I could also remove the try-catch block so the exception bubbles up

**Comment by ChristophWurst** [comment]
Every time we add more complexity to ``OC\User\Session``… God kills a kitten. This class cries for refactoring…

Unit tests added and try-catch block removed. @PVince81 @rullzer @nickvergessen please (re-)review

**Comment by PVince81** [comment]
:+1:

**Comment by guruz** [comment]
Code looks OK 👍

**Comment by PVince81** [comment]
@ChristophWurst mind rebasing to get a bit more greenery ?

**Comment by ChristophWurst** [comment]
@PVince81 done

**Comment by nickvergessen** [comment]
When installing via occ it fails:
```
10:27:34   [OCP\Session\Exceptions\SessionNotAvailableException]  
10:27:34   Memory session does not have an ID                     
10:27:34                                                          
10:27:34 
10:27:34 Exception trace:
10:27:34  () at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/lib/private/Session/Memory.php:102
10:27:34  OC\Session\Memory->getId() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/lib/private/User/Session.php:472
10:27:34  OC\User\Session->createSessionToken() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/lib/private/Setup.php:376
10:27:34  OC\Setup->install() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/core/Command/Maintenance/Install.php:84
10:27:34  OC\Core\Command\Maintenance\Install->execute() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/3rdparty/symfony/console/Command/Command.php:259
10:27:34  Symfony\Component\Console\Command\Command->run() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/3rdparty/symfony/console/Application.php:844
10:27:34  Symfony\Component\Console\Application->doRunCommand() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/3rdparty/symfony/console/Application.php:192
10:27:34  Symfony\Component\Console\Application->doRun() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/3rdparty/symfony/console/Application.php:123
10:27:34  Symfony\Component\Console\Application->run() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/lib/private/Console/Application.php:145
10:27:34  OC\Console\Application->run() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/console.php:94
10:27:34  require_once() at /ssd/jenkins/workspace/server-master-linux-php7-ci@4/database/sqlite/label/SLAVE/occ:11
```

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -470,11 +470,39 @@ public function createSessionToken(IRequest $request, $uid, $loginName, $passwor
 		$name = isset($request->server['HTTP_USER_AGENT']) ? $request->server['HTTP_USER_AGENT'] : 'unknown browser';
 		try {
 			$sessionId = $this->session->getId();
-			$this->tokenProvider->generateToken($sessionId, $uid, $loginName, $password, $name);
+			$pwd = $this->getPassword($password);
+			$this->tokenProvider->generateToken($sessionId, $uid, $loginName, $pwd, $name);
+			return true;
 		} catch (SessionNotAvailableException $ex) {
+			// This can happen with OCC, where a memory session is used
+			// if a memory session is used, we shouldn't create a session token anyway
```
Comment (ChristophWurst) [inline lib/private/User/Session.php]: @nickvergessen I added the try-catch block again as it actually makes sense to catch the exception here.

**Comment by PVince81** [comment]
Tested, works :+1: 

@nickvergessen can you redo a quick test since you were having issues ?

**Comment by nickvergessen** [comment]
That was jenkins, not me in particular.

But tested it quickly, sync client seems to still work, so fine by me :+1:

**Comment by ChristophWurst** [comment]
@dragotin can we move the discussion from https://github.com/owncloud/core/commit/ec929f07f21341ed0e679ca27dc7e3cacd8b1b2d#commitcomment-17832717 here? Otherwise nobody will ever find that again. Thanks

**Comment by dragotin** [comment]
@ChristophWurst don't worry, I just went through this PR and gave my completely clueless comments. No particular reason :-)

The only thing that I would not agree on is the error handling. There has to be any controlled way to deal with the error. If it means "we do nothing special", ok, fair enough, but here it looks like it is just not handled. 

I did not know that there can't be other implementations of the TokenProvider. 

Thanks anyways :)

**Pull request merged**